diff --git hardware/msm7k/Android.mk hardware/msm7k/Android.mk
index 6b916d9..a14bf02 100644
--- hardware/msm7k/Android.mk
+++ hardware/msm7k/Android.mk
@@ -18,7 +18,8 @@ common_msm_dirs := liblights librpc libgralloc-qsd8k libstagefrighthw
 msm7k_dirs := $(common_msm_dirs) boot libaudio libcopybit
 qsd8k_dirs := $(common_msm_dirs) libaudio-qsd8k dspcrashd libcopybit
 msm7x30_dirs := $(common_msm_dirs) libaudio-msm7x30 liboverlay libcopybit libsensors
-msm8660_dirs := $(common_msm_dirs) libaudio-msm8660 liboverlay libcopybit dspcrashd
+msm8660_dirs := $(common_msm_dirs) liboverlay libcopybit dspcrashd
+#msm8660_dirs := $(common_msm_dirs) libaudio-msm8660 liboverlay libcopybit dspcrashd
 
 ifeq ($(TARGET_BOARD_PLATFORM),msm7k)
   ifeq "$(findstring msm7630,$(TARGET_PRODUCT))" "msm7630"
