Index: 89ed6bd..3bf3254 100644
--- dalvik/vm/Dvm.mk
+++ dalvik/vm/Dvm.mk
@@ -287,7 +287,7 @@ ifeq ($(dvm_arch),arm)
   endif
 
   ifeq ($(BOARD_USES_QCOM_HARDWARE),true)
-    LOCAL_WHOLE_STATIC_LIBRARIES += libqc-dalvik
+    #LOCAL_WHOLE_STATIC_LIBRARIES += libqc-dalvik
   endif
 endif

Index: 8ac2165..e9ce964 100644
--- external/sqlite/dist/Android.mk
+++ external/sqlite/dist/Android.mk
@@ -40,7 +40,7 @@ LOCAL_SHARED_LIBRARIES += liblog \
 LOCAL_WHOLE_STATIC_LIBRARIES := libsqlite3_android
 
 ifeq ($(BOARD_USES_QCOM_HARDWARE),true)
-LOCAL_WHOLE_STATIC_LIBRARIES += libqc-sqlite
+#LOCAL_WHOLE_STATIC_LIBRARIES += libqc-sqlite
 LOCAL_SHARED_LIBRARIES += libcutils
 endif
 
diff --git a/Android.mk b/Android.mk
index 59a8aa3..cac36fe 100644
--- external/skia/Android.mk
+++ external/skia/Android.mk
@@ -234,7 +234,7 @@ LOCAL_STATIC_LIBRARIES := \
 	libgif
 
 ifeq ($(BOARD_USES_QCOM_HARDWARE),true)
-	LOCAL_WHOLE_STATIC_LIBRARIES += libqc-skia
+	#LOCAL_WHOLE_STATIC_LIBRARIES += libqc-skia
 endif
 
 
